File: src/main/java/network/crypta/launcher/ThemeSwitcher.java
package network.crypta.launcher;

import com.formdev.flatlaf.FlatDarkLaf;
import com.formdev.flatlaf.FlatLaf;
import com.formdev.flatlaf.FlatLightLaf;
import com.formdev.flatlaf.themes.FlatMacDarkLaf;
import com.formdev.flatlaf.themes.FlatMacLightLaf;
import com.jthemedetecor.OsThemeDetector;
import java.awt.Font;
import java.util.function.Consumer;
import javax.swing.JDialog;
import javax.swing.JFrame;
import javax.swing.LookAndFeel;
import javax.swing.SwingUtilities;
import javax.swing.UIManager;
import javax.swing.plaf.FontUIResource;
import network.crypta.fs.AppEnv;

import static network.crypta.launcher.LauncherLog.logDebug;
import static network.crypta.launcher.LauncherLog.logWarn;

/**
 * Applies and maintains FlatLaf theme selection based on the current operating-system appearance.
 *
 * <p>This class bridges OS theme detection with Swing look-and-feel updates for the launcher UI. It
 * performs initial one-time setup, selects a platform-appropriate FlatLaf variant, and registers a
 * listener so theme changes propagate while the application is running. Updates are marshaled onto
 * the Swing event dispatch thread to keep the UI state consistent with Swing threading rules. The
 * implementation also includes platform nuances, such as macOS appearance hints and Flatpak-aware
 * native decoration toggles.
 *
 * <p>Key behaviors:
 *
 * <ul>
 *   <li>Uses portal-aware theme detection when available, with fallback detector behavior.
 *   <li>Prefers synchronous initial LAF application to avoid startup flicker.
 *   <li>Supports clean listener shutdown to prevent stale callback retention.
 * </ul>
 */
public final class ThemeSwitcher {
  private static OsThemeDetector detector;
  private static Consumer<Boolean> listener;

  private ThemeSwitcher() {}

  /**
   * Installs launcher look-and-feel based on the current OS theme and registers change tracking.
   *
   * <p>The initial theme is applied immediately, then an OS theme listener is attached, so the
   * following dark/light transitions update the active look-and-feel.
   */
  public static synchronized void install() {
    try {
      System.setProperty("apple.awt.application.appearance", "system");
    } catch (Exception e) {
      logDebug("Failed to set macOS appearance system property", e);
    }

    try {
      if (new AppEnv().isLinux()) {
        JFrame.setDefaultLookAndFeelDecorated(true);
        JDialog.setDefaultLookAndFeelDecorated(true);
      }
    } catch (Exception e) {
      logDebug("Failed to enable client decorations on Linux", e);
    }

    OsThemeDetector osThemeDetector = FlatpakAwareOsThemeDetector.getDetector();
    detector = osThemeDetector;
    applyFor(osThemeDetector.isDark(), true);

    Consumer<Boolean> newListener = isDark -> applyFor(Boolean.TRUE.equals(isDark), false);
    listener = newListener;
    osThemeDetector.registerListener(newListener);
  }

  /**
   * Unregisters the active OS theme listener and clears retained detector references.
   *
   * <p>This method is safe to call multiple times and is intended for orderly launcher shutdown.
   */
  public static synchronized void shutdown() {
    if (detector != null && listener != null) {
      detector.removeListener(listener);
      if (detector instanceof AutoCloseable closeable) {
        try {
          closeable.close();
        } catch (Exception e) {
          logDebug("Failed to close OS theme detector", e);
        }
      }
    }
    detector = null;
    listener = null;
  }

  private static void applyFor(boolean useDark, boolean synchronous) {
    boolean mac = new AppEnv().isMac();
    LookAndFeel laf = createLaf(useDark, mac);
    applyLaf(laf, mac, synchronous);
  }

  private static LookAndFeel createLaf(boolean useDark, boolean mac) {
    if (mac) {
      return useDark ? new FlatMacDarkLaf() : new FlatMacLightLaf();
    }
    return useDark ? new FlatDarkLaf() : new FlatLightLaf();
  }

  private static void applyLaf(LookAndFeel laf, boolean mac, boolean synchronous) {
    Runnable apply = () -> applyLafInternal(laf, mac, synchronous);
    if (synchronous) {
      if (SwingUtilities.isEventDispatchThread()) {
        apply.run();
      } else {
        try {
          SwingUtilities.invokeAndWait(apply);
        } catch (InterruptedException e) {
          Thread.currentThread().interrupt();
          logDebug("invokeAndWait failed; applying LAF on caller thread", e);
          apply.run();
        } catch (Exception e) {
          logDebug("invokeAndWait failed; applying LAF on caller thread", e);
          apply.run();
        }
      }
      return;
    }
    SwingUtilities.invokeLater(apply);
  }

  private static void applyLafInternal(LookAndFeel laf, boolean mac, boolean synchronous) {
    try {
      if (mac) {
        UIManager.put("defaultFont", new FontUIResource("SansSerif", Font.PLAIN, 13));
      }
      boolean inFlatpak = !System.getenv().getOrDefault("FLATPAK_ID", "").isEmpty();
      FlatLaf.setUseNativeWindowDecorations(!mac && !inFlatpak);
      UIManager.setLookAndFeel(laf);
      System.setProperty("swing.defaultlaf", laf.getClass().getName());
      FlatLaf.setup(laf);
      if (!synchronous) {
        FlatLaf.updateUI();
      }
    } catch (Exception e) {
      logWarn("Failed to apply FlatLaf theme", e);
    }
  }
}

File: src/main/kotlin/network/crypta/launcher/ThemeSwitcher.kt
package network.crypta.launcher

import com.formdev.flatlaf.FlatDarkLaf
import com.formdev.flatlaf.FlatLaf
import com.formdev.flatlaf.FlatLightLaf
import com.formdev.flatlaf.themes.FlatMacDarkLaf
import com.formdev.flatlaf.themes.FlatMacLightLaf
import com.jthemedetecor.OsThemeDetector
import java.util.function.Consumer
import javax.swing.LookAndFeel
import javax.swing.SwingUtilities
import javax.swing.UIManager
import javax.swing.plaf.FontUIResource
import network.crypta.fs.AppEnv

/**
 * Applies a FlatLaf look-and-feel that tracks the operating system theme and keeps Swing components
 * synchronized as the OS preference changes.
 *
 * This singleton is designed for early application startup: call [install] before any Swing
 * components are created so the initial look-and-feel is consistent and does not flash from a
 * default theme. It selects a platform-appropriate FlatLaf variant, configures window decorations
 * for Linux/Flatpak, and wires an [OsThemeDetector] listener so later OS theme changes update the
 * UI. The implementation keeps the minimal mutable state needed to unregister the listener in
 * [shutdown], and it is safe to call from any thread because internal dispatch uses the EDT.
 *
 * Notable behaviors:
 * <ul>
 * <li>Initial theme application is synchronous to avoid UI creation races.</li>
 * <li>Live switches refresh the UI on the EDT.</li>
 * <li>macOS applies a default logical font to avoid JDK-8355079.</li>
 * </ul>
 *
 * @see FlatpakAwareOsThemeDetector
 * @see FlatLaf
 * @see AppEnv
 */
object ThemeSwitcher {
  private class ThemeDetectionState {
    @Volatile var detector: OsThemeDetector? = null

    @Volatile var listener: Consumer<Boolean>? = null
  }

  private val state = ThemeDetectionState()

  /**
   * Initialize OS theme detection and apply the matching FlatLaf before UI creation.
   *
   * Call this once during application startup, before any Swing component is constructed, so the
   * initial theme is applied synchronously on the EDT and remains stable. The method configures
   * platform-specific details (macOS appearance hint and Linux client decorations), creates a
   * detector via [FlatpakAwareOsThemeDetector], applies the current theme immediately, and then
   * registers a listener for later OS theme changes. Repeated calls replace the stored detector
   * reference, but the intent is a single early call to avoid redundant listeners.
   */
  fun install() {
    // macOS: ensure the system appearance is reported to Java/Swing
    try {
      System.setProperty("apple.awt.application.appearance", "system")
    } catch (e: Exception) {
      logDebug("Failed to set macOS appearance system property", e)
    }
    try {
      if (AppEnv().isLinux()) {
        javax.swing.JFrame.setDefaultLookAndFeelDecorated(true)
        javax.swing.JDialog.setDefaultLookAndFeelDecorated(true)
      }
    } catch (t: Throwable) {
      logDebug("Failed to enable client decorations on Linux/Flatpak", t)
    }
    val det = FlatpakAwareOsThemeDetector.getDetector()
    state.detector = det

    // Apply the current theme synchronously (must happen before any Swing components are created)
    val initialDark = det.isDark
    applyFor(initialDark, synchronous = true)

    // Listen for changes and switch LAF live
    val consumer = Consumer<Boolean> { isDark -> applyFor(isDark) }
    state.listener = consumer
    det.registerListener(consumer)
  }

  /**
   * Stop OS theme change reporting and release any registered listener.
   *
   * This is intended for application shutdown to avoid leaks or spurious callbacks after the UI is
   * torn down. If [install] was never called, this method is a no-op. It clears the cached detector
   * and listener references after attempting to unregister them, and it does not touch the current
   * look-and-feel state.
   */
  fun shutdown() {
    val det = state.detector
    val c = state.listener
    if (det != null && c != null) det.removeListener(c)
    state.detector = null
    state.listener = null
  }

  private fun applyFor(useDark: Boolean, synchronous: Boolean = false) {
    val mac = AppEnv().isMac()
    val laf = createLaf(useDark, mac)
    applyLaf(laf, mac, synchronous)
  }

  private fun createLaf(useDark: Boolean, mac: Boolean): LookAndFeel =
    if (mac) {
      if (useDark) FlatMacDarkLaf() else FlatMacLightLaf()
    } else {
      if (useDark) FlatDarkLaf() else FlatLightLaf()
    }

  private fun applyLaf(laf: LookAndFeel, mac: Boolean, synchronous: Boolean) {
    val apply: () -> Unit = { applyLafInternal(laf, mac, synchronous) }

    if (synchronous) {
      if (SwingUtilities.isEventDispatchThread()) {
        apply()
      } else {
        try {
          SwingUtilities.invokeAndWait(apply)
        } catch (e: Exception) {
          logDebug("invokeAndWait failed; applying LAF synchronously on caller thread", e)
          apply()
        }
      }
    } else {
      SwingUtilities.invokeLater(apply)
    }
  }

  private fun applyLafInternal(laf: LookAndFeel, mac: Boolean, synchronous: Boolean) {
    try {
      if (mac) {
        // Work around JDK-8355079: set a sane default logical font on macOS
        UIManager.put("defaultFont", FontUIResource("SansSerif", java.awt.Font.PLAIN, 13))
      }
      configureWindowDecorations()
      applyLookAndFeel(laf)
      setDefaultLafProperty(laf)
      FlatLaf.setup(laf)
      // For live switches after startup, refresh UI
      if (!synchronous) FlatLaf.updateUI()
    } catch (e: Exception) {
      logWarn("Failed to apply FlatLaf theme", e)
    }
  }

  private fun configureWindowDecorations() {
    // Use FlatLaf client decorations on Linux/Flatpak to avoid light server-side title bars
    val inFlatpak = System.getenv("FLATPAK_ID")?.isNotEmpty() == true
    val useNativeDeco = AppEnv().isMac().not() && !inFlatpak
    FlatLaf.setUseNativeWindowDecorations(useNativeDeco)
  }

  private fun applyLookAndFeel(laf: LookAndFeel) {
    // Apply LAF explicitly via UIManager to catch errors, then let FlatLaf do extra setup
    try {
      UIManager.setLookAndFeel(laf)
    } catch (t: Throwable) {
      logDebug("UIManager.setLookAndFeel() failed", t)
    }
  }

  private fun setDefaultLafProperty(laf: LookAndFeel) {
    // Also set swing.defaultlaf so the EDT initialization cannot override us to GTK
    try {
      System.setProperty("swing.defaultlaf", laf.javaClass.name)
    } catch (t: Throwable) {
      logDebug("Failed to set swing.defaultlaf system property", t)
    }
  }
}
